Encryption.md

```diff
@@ -132,49 +203,164 @@ the encryption buffer is comprised of an IV, ciphertext and tag, as described in
+
+  /** Column metadata for this chunk. This is the same content as what is at
+   * file_path/file_offset.  Having it here has it replicated in the file
+   * metadata.
+   **/
+  3: optional ColumnMetaData meta_data
+..
+}
+```
+
 ### Encrypted footer mode
 
 In files with sensitive column data, a good security practice is to encrypt not only the 
-secret columns, but also the file footer metadata, with a separate footer key. This hides
+secret columns, but also the file footer metadata with a separate footer key. This hides
 the file schema / column names, number of rows, key-value properties, column sort order, 
 column data offset and size, list of encrypted columns and metadata of the column encryption keys. 
-It also makes the footer tamper-proof.
+It also makes the footer tamper-proof and helps to prevent module replacement attacks.
 
-The columns encrypted with the same key as the footer, don't split the ColumnMetaData from the 
+The columns encrypted with the same key as the footer don't separate the ColumnMetaData from the 
 ColumnChunks, leaving it at the original location, `optional ColumnMetaData meta_data`. This field
-is not set for columns enrypted with a column-specific key.
+is not set for columns encrypted with a column-specific key.
 
 A Thrift-serialized `FileCryptoMetaData` structure is written after the footer. It contains 
 information on the file encryption algorithm and on the footer (offset in 
 the file, and an optional key metadata, with a maximal length of 512). Then 
-the length of this structure is written, as a 4-byte little endian integer. Then a final 
+the length of this structure is written as a 4-byte little endian integer, followed by a final 
 magic string, "PARE".
-
-Only the `FileCryptoMetaData` is written as a plaintext, all other file parts are protected
+The same magic bytes are written at the beginning of the file (offset 0).
+Parquet readers start file parsing by reading and checking the magic string. Therefore, the 
+encrypted footer mode uses a new magic string ("PARE") in order to  instruct readers to look 
+for a file crypto metadata at the end of the file - and also to immediately inform 
+legacy readers (expecting ‘PAR1’ bytes) that they can’t parse this file.
+
+```c
+/** Crypto metadata for files with encrypted footer **/
+struct FileCryptoMetaData {
+  /** 
+   * Encryption algorithm. Note that this field is only used for files
+   * with encrypted footer. Files with plaintext footer store the algorithm id
+   * inside footer (FileMetaData structure).
+   */
+  1: required EncryptionAlgorithm encryption_algorithm
+    
+  /** Retrieval metadata of key used for encryption of footer, 
+   *  and (possibly) columns **/
+  2: optional binary footer_key_metadata
+
+  /** Offset of encrypted Parquet footer **/
+  3: required i64 footer_offset
+}
+```
+
+Only the `FileCryptoMetaData` is written as a plaintext; all other file parts are protected
 (as needed) with appropriate keys.
```

Comment:
I think it's a bad idea to use "(as needed)". All encrypted structures should be specified in this doc.